src/main/java/de/teamlapen/vampirism/blocks/TentBlock.java

```diff
@@ -1,102 +1,293 @@
+            return true;
+        }
+        IForgeDimension.SleepResult sleepResult = world.getDimension().canSleepAt(playerEntity,blockPos);
+        if (sleepResult != IForgeDimension.SleepResult.BED_EXPLODES) {
+            if (sleepResult == IForgeDimension.SleepResult.DENY) return true;
+            if (blockState.get(BedBlock.OCCUPIED)) {
+                playerEntity.sendStatusMessage(new TranslationTextComponent("text.vampirism.tent.occupied"), true);
+                return true;
+            } else {
+                playerEntity.trySleep(blockPos).ifLeft(sleepResult1 -> {
+                    if (sleepResult1 != null) {
+                        playerEntity.sendStatusMessage(sleepResults.getOrDefault(sleepResult1, sleepResult1.getMessage()), true);
+                    }
+                }).ifRight(u -> {
+                    this.setBedOccupied(blockState, world, blockPos, null, true);
+                    setTentSleepPosition(playerEntity, blockPos, playerEntity.world.getBlockState(blockPos).get(POSITION), playerEntity.world.getBlockState(blockPos).get(HORIZONTAL_FACING));
+                });
+                return true;
+            }
+        } else {
+            world.removeBlock(blockPos, false);
+            BlockPos blockPos1 = blockPos.offset(blockState.get(HORIZONTAL_FACING).getOpposite());
+            if (world.getBlockState(blockPos1).getBlock() == this) {
+                world.removeBlock(blockPos1, false);
+            }
+            world.createExplosion(null, DamageSource.netherBedExplosion(), (double) blockPos.getX() + 0.5D, (double) blockPos.getY() + 0.5D, (double) blockPos.getZ() + 0.5D, 5.0F, true, Explosion.Mode.DESTROY);
+            return true;
+        }
+    }
+
+    @Override
+    public void setBedOccupied(BlockState state, IWorldReader world, BlockPos pos, LivingEntity sleeper, boolean occupied) {
+        if(world instanceof IWorldWriter){
+            forWholeTent(pos, state, ((direction, blockPos) -> {
+                BlockState blockState = world.getBlockState(blockPos);
+                if(blockState.getBlock() instanceof TentBlock){
+                    ((IWorldWriter & IWorldReader)world).setBlockState(pos, world.getBlockState(pos).with(BlockStateProperties.OCCUPIED, occupied), 2);
+                }
+            }));
+        }
+    }
+
+    @Override
+    public boolean isBed(BlockState state, IBlockReader world, BlockPos pos, @Nullable Entity player) {
+        return true;
+    }
+
+    @Override
+    public Direction getBedDirection(BlockState state, IWorldReader world, BlockPos pos) {
+        switch (state.get(POSITION)){
+            case 0:
+            case 3:
+                return state.get(HORIZONTAL_FACING).getOpposite();
+            default:
+                return state.get(HORIZONTAL_FACING);
+        }
+    }
+
+    @Override
+    public void onReplaced(BlockState state, @Nonnull World world, @Nonnull BlockPos pos, @Nonnull BlockState newState, boolean isMoving) {
```

Comment:
While you are at it, this should probably be migrated to #onBlockHarvested like BedBlock does and also check if the blocks we are destroying are actually tent blocks